pkg/apis/core/types.go

```diff
@@ -1670,6 +1673,29 @@ type CSIVolumeSource struct {
 	NodePublishSecretRef *LocalObjectReference
 }
 
+// EphemeralVolumeSource represents ephemeral storage that is handled by a normal storage driver.
+type EphemeralVolumeSource struct {
+	// Will be created as a stand-alone PVC to provision the volume.
+	//
+	// The pod in which this EphemeralVolumeSource is embedded will be the
+	// owner, i.e. the PVC will be deleted together with the pod.
+	// The name of the PVC will be `<pod name>-<volume name>` where `<volume name>`
+	// is the name from the `PodSpec.Volumes` array entry. Care must
+	// be taken to keep this concatendated name shorter enough that it
+	// can be used as object name. Labels will be copied from the pod.
+	//
+	// This field is read-only and no changes will be made by Kubernetes
+	// to the PVC after it has been created.
+	//
+	// Required, must not be nil.
+	VolumeClaim *PersistentVolumeClaimSpec
```

Comment:
In the KEP, I proposed to embed a full `PersistentVolumeClaim` (= `ObjectMeta` + `PersistentVolumeClaimSpec` + redundant `PersistentVolumeClaimStatus`, as in `StatefulSet`) which then led to a new `PersistentVolumeClaimTemplate` (= `ObjectMeta` + `PersistentVolumeClaimSpec`). My thinking at that time was that it would be useful to grant users full explicit control over the labels that get added to the created PVC.

However, that flexibility also has downsides, like the redundant name in the `ObjectMeta`. I then realized that tools like "kustomize" would have to be updated to deal correctly with explicit labels: currently, kustomize's `commonLabels` feature only adds labels to the meta data of types it knows about and doesn't add them to an embedded `PersistentVolumeClaim`.

Therefore I'd like to propose that we switch to one of the proposals during the KEP review and just embed a `PersistentVolumeClaimSpec`, with labels being copied from the pod meta data (implicit approach).

I'm open whether that should be *all* labels or whether we only copy certain labels, like those with the `app.kubernetes.io/` prefix (https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels).

@saad-ali: what do you think?

Reply:
@jsafrane already said that he prefers `PersistentVolumeTemplate` (= `ObjectMeta` + `PersistentVolumeClaimSpec`). Let me try to get to the bottom of what that means for the API:
- name, generatename and namespace are redundant
- all of the generated fields (UID, ResourceVersion, Generation, CreationTimestamp, DeletionTimestamp, DeletionGracePeriodSeconds) are redundant
- ownerreference is superseded or at least modified: does it make sense to allow more than one owner? I asked on Slack what multiple owners mean, but never fully understood that.
- Finalizers and ManagedFields don't make sense.
- The only fields which make sense are labels and annotations.

If we compose `PersistentVolumeTemplate` of `ObjectMeta`, should the user be allowed to set all fields, including those that don't make sense? I think the `StatefulSet` allows it. That may be a historic accident; we could decide to tighten validation such that for `PersistentVolumeClaimTemplate`, *only* labels and annotations can be set.

@jsafrane: agreed?

Reply:
I checked how Pod templates are validated in Deployments / StatefulSets etc, they do just very generic validation:

https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/pkg/apis/core/validation/validation.go#L4409-L4420

I think we should do the same - copy validation from PVC where it makes sense and don't restrict UID / Generation / CreationTimestamp / etc. The controller should only overwrite things it needs to control (GenerateName, Name).

---

BTW, I think we discussed this and you convinced me that we should use the same approach as StatefulSet, i.e. whole PVC object (incl. status) as a template:

https://github.com/kubernetes/kubernetes/blob/171b486c3483d763a06aa67ead4ff6e93d23874e/pkg/apis/apps/types.go#L130

I looked at how it's validated, but [it is not validated at all](https://github.com/kubernetes/kubernetes/blob/905e7510c84676fe5c6428f1038996e099a30f68/pkg/apis/apps/validation/validation.go#L71) (ehm...)

I see you current proposal, ClaimTemplate with ObjectMeta + Spec better.

Reply:
Okay, so to recap:
- we add `PersistentVolumeClaimTemplate` = `ObjectMeta` + `PersistentVolumeClaimSpec`
- only labels and annotations are copied by the controller, the rest of `ObjectMeta` is ignored
- validation is reused where it exists, but no further restrictions are enforced regarding the unused fields

Works for me. I'll extend the PR accordingly, using separate commits, so if we want to go back, we can.

Reply:
> I checked how Pod templates are validated in Deployments / StatefulSets etc, they do just very generic validation

statefulsets are under-validated (xref https://github.com/kubernetes/kubernetes/issues/41153, https://github.com/kubernetes/kubernetes/issues/84698, https://github.com/kubernetes/kubernetes/issues/81147, https://github.com/kubernetes/kubernetes/issues/60619, https://github.com/kubernetes/kubernetes/issues/51306), let's not repeat those mistakes. I would expect us to reuse existing validation to ensure impossible-to-succeed things are not created.

> only labels and annotations are copied by the controller, the rest of ObjectMeta is ignored

persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored.

> we add PersistentVolumeClaimTemplate = ObjectMeta + PersistentVolumeClaimSpec

what does this do to the size of the podspec schema? importing a complex type definition for spawning new objects into what has been our leaf object is a pretty notable change.

Reply:
> persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored

We can specify that all other fields must be empty and I can add validation for that. However, that then leads to the question whether ObjectMeta is the right type to use in the first place.

`PersistentVolumeClaimTemplate` could also be defined as:
```
type PersistentVolumeClaimTemplate struct {
    Labels map[string]string
    Annotations map[string]string
    VolumeSpec PersistentVolumeClaimSpec
}
```

> what does this do to the size of the podspec schema?

I have pushed the change into the branch for this PR, so we can check that. Where do I find the podspec schema?

Reply:
> I'll extend the PR accordingly, using separate commits, so if we want to go back, we can.

I ended up archiving the previous source and just squashed my changes so (currently) 
77ad5d8 has the API as originally proposed in the KEP. But the defaulting test fails, so I'll have to revise that a bit.

Reply:
> persisting and ignoring data is misleading. I would expect to only allow persisting data in fields that will be honored.

We don'd to that for PodTemplates - validation allows UID, Generation, Name and maybe other stuff that's thrown away to be set there.

Reply:
> We don'd to that for PodTemplates - validation allows UID, Generation, Name and maybe other stuff that's thrown away to be set there.

Yes, and that causes confusion and bug reports.

Reply:
I agree with @liggitt, the existing behavior isn't a good example to follow and I'd be happy to do something better - we just need to agree *what* that is. Embedding just `PersistentVolumeClaimSpec` was my attempt to avoid that entire source of confusion, but then there's no place to specify labels and annotations.

@liggitt : do you have some preferred solution?

To recap the alternatives:
- embed `PersistentVolumeClaimSpec`, labels (potentially limited to apps.kubernetes.io/*) copied from pod
- embed `ObjectMeta` + `PersistentVolumeClaimSpec`, with validation that limits `ObjectMeta` to labels and annotations
- custom struct as in https://github.com/kubernetes/kubernetes/pull/91680#discussion_r436014707

Reply:
> edit: I was thinking of the PV type, not the PVC type, when I was concerned about complexity. 

That's simple then: PV does not change at all, nor does the code which handles provisioning for the PVC.

> However, the questions about spawning N objects from pod creation still remain.

That's a numbers game. How large is N and how many pods will use several inline volumes? We don't know yet and it may entirely depend on how quick provisioning is in practice, so we have to build something, ask people to try it out and gather feedback.

But even if it turns out to not scale well, it should still be useful for long-lived pods with few replicas (the memcached example).

Reply:
I pushed a version with strict validation: only labels and annotations are allowed in the embedded ObjectMeta.

While looking for a way to do this via whitelisting those fields (less error-prone than blacklisting), I found a precedent and used a generalized version of that code: https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/pkg/apis/core/validation/validation.go#L2649-L2666

Reply:
Another data point: if we go with embedded labels + annotations, "kustomize" needs a slight update, otherwise it does not apply common labels and annotations to the new fields. It does not matter whether we use ObjectMeta or a custom struct. On the bright side, I looked into this and have a patch ready for kustomize if we decide to add this API.

Reply:
We decided to embed `ObjectMeta` + `PersistentVolumeClaimSpec`, so I consider this resolved.